### PR TITLE
SONAR-8004 WS settings/set handles property set

### PR DESCRIPTION
**Pull request by teryk**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java [hunk 1]
```diff
@@ -203,6 +292,22 @@ private static SetRequest toWsRequest(Request request) {
...
+  private static List<Map<String, String>> readFieldsValues(Request request) {
+    Type type = new TypeToken<Map<String, String>>() {
+    }.getType();
+    Gson gson = GsonHelper.create();
+
+    return request.multiParam(PARAM_FIELDS_VALUES).stream()
+      .map(jsonString -> {
+        try {
+          return (Map<String, String>) gson.fromJson(jsonString, type);
+        } catch (JsonParseException e) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: MAJOR Either log or rethrow this exception. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1166)

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java [hunk 2]
```diff
@@ -126,6 +145,67 @@ public void handle(Request request, Response response) throws Exception {
...
+    } else {
+      validate(request, component);
+      dbClient.propertiesDao().insertProperty(dbSession, toProperty(request, component));
+      if (!component.isPresent()) {
+        settingsChangeNotifier.onGlobalPropertyChange(request.getKey(), persistedValue(request));
+      }
+    }
+  }
+
+  private void doHandlePropertySet(DbSession dbSession, SetRequest request, @Nullable PropertyDefinition definition, Optional<ComponentDto> component) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: MINOR Specify a "ComponentDto" parameter instead. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java [hunk 3]
```diff
@@ -126,6 +145,67 @@ public void handle(Request request, Response response) throws Exception {
...
+    validatePropertySet(request, definition, component);
+
+    Set<String> fieldKeys = definition.fields().stream()
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: MAJOR NullPointerException might be thrown as 'definition' is nullable here [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2259)

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java [hunk 4]
```diff
@@ -126,6 +145,67 @@ public void handle(Request request, Response response) throws Exception {
...
+    dbClient.propertiesDao().insertProperty(dbSession, new PropertyDto().setKey(key).setValue(inlinedFieldKeys).setResourceId(componentId));
+
+    List<Map<String, String>> fieldsValues = request.getFieldsValues();
+    IntStream.rangeClosed(1, fieldsValues.size())
+      .boxed()
+      .flatMap(i -> fieldsValues.get(i - 1).entrySet().stream().map(entry -> new KeyValue(key + "." + i + "." + entry.getKey(), entry.getValue())))
+      .forEach(keyValue -> dbClient.propertiesDao().insertProperty(dbSession, new PropertyDto().setKey(keyValue.key).setValue(keyValue.value).setResourceId(componentId)));
+  }
+
+  private void validatePropertySet(SetRequest request, @Nullable PropertyDefinition definition, Optional<ComponentDto> component) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: MINOR Specify a "ComponentDto" parameter instead. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java [hunk 5]
```diff
@@ -126,6 +145,67 @@ public void handle(Request request, Response response) throws Exception {
...
+    checkRequest(definition != null, "Setting '%s' is undefined", request.getKey());
+    checkRequest(PropertyType.PROPERTY_SET.equals(definition.type()), "Parameter '%s' is used for setting of property set type only", PARAM_FIELDS_VALUES);
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: MAJOR NullPointerException might be thrown as 'definition' is nullable here [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2259)

File: sonar-ws/src/main/java/org/sonarqube/ws/client/setting/SettingsService.java [hunk 6]
```diff
@@ -61,6 +63,9 @@ public void set(SetRequest request) {
     call(new PostRequest(path(ACTION_SET))
       .setParam(PARAM_KEY, request.getKey())
       .setParam(PARAM_VALUE, request.getValue())
+      // TODO doesn't work for multi param
```
Comment (SonarTech) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/setting/SettingsService.java]: INFO Complete the task associated to this TODO comment. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues
* MAJOR 3 major
* MINOR 2 minor
* INFO 1 info

Watch the comments in this conversation to review them.

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java [hunk 7]
```diff
@@ -126,6 +146,64 @@ public void handle(Request request, Response response) throws Exception {
...
+    if (component.isPresent()) {
+      settingsUpdater.deleteComponentSetting(dbSession, key, component.get());
+    } else {
+      settingsUpdater.deleteGlobalSetting(dbSession, key);
+    }
+  }
+
+  private void validatePropertySet(SetRequest request, @Nullable PropertyDefinition definition, Optional<ComponentDto> component) {
+    checkRequest(definition != null, "Setting '%s' is undefined", request.getKey());
+    checkRequest(PropertyType.PROPERTY_SET.equals(definition.type()), "Parameter '%s' is used for setting of property set type only", PARAM_FIELD_VALUES);
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: MAJOR NullPointerException might be thrown as 'definition' is nullable here [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2259)

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java [hunk 8]
```diff
@@ -126,6 +146,64 @@ public void handle(Request request, Response response) throws Exception {
...
+    deleteSettings(dbSession, component, key);
+    dbClient.propertiesDao().insertProperty(dbSession, new PropertyDto().setKey(key).setValue(inlinedFieldKeys).setResourceId(componentId));
+
+    List<Map<String, String>> fieldValues = request.getFieldValues();
+    IntStream.of(fieldIds).boxed()
+      .flatMap(i -> fieldValues.get(i - 1).entrySet().stream().map(entry -> new KeyValue(key + "." + i + "." + entry.getKey(), entry.getValue())))
+      .forEach(keyValue -> dbClient.propertiesDao().insertProperty(dbSession, toFieldProperty(keyValue, componentId)));
+  }
+
+  private void deleteSettings(DbSession dbSession, Optional<ComponentDto> component, String key) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: MINOR Specify a "ComponentDto" parameter instead. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java [hunk 9]
```diff
@@ -138,6 +216,17 @@ private void validate(SetRequest request, Optional<ComponentDto> component) {
     checkComponentQualifier(request, definition, component);
   }
 
+  private void checkFieldType(SetRequest request, @Nullable PropertyDefinition definition, ListMultimap<String, String> valuesByFieldKeys) {
+    for (PropertyFieldDefinition fieldDefinition : definition.fields()) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: MAJOR NullPointerException might be thrown as 'definition' is nullable here [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2259)

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java [hunk 10]
```diff
@@ -203,6 +301,24 @@ private static SetRequest toWsRequest(Request request) {
...
+      .map(json -> readOneFieldValues(json, key))
+      .collect(Collectors.toList());
+  }
+
+  private static Map<String, String> readOneFieldValues(String json, String key) {
+    Type type = new TypeToken<Map<String, String>>() {
+    }.getType();
+    try {
+      return (Map<String, String>) GsonHelper.create().fromJson(json, type);
+    } catch (JsonParseException e) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: MAJOR Either log or rethrow this exception. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1166)

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java [hunk 11]
```diff
@@ -126,6 +146,64 @@ public void handle(Request request, Response response) throws Exception {
...
+
+  private void deleteSettings(DbSession dbSession, Optional<ComponentDto> component, String key) {
+    if (component.isPresent()) {
+      settingsUpdater.deleteComponentSetting(dbSession, key, component.get());
+    } else {
+      settingsUpdater.deleteGlobalSetting(dbSession, key);
+    }
+  }
+
+  private void validatePropertySet(SetRequest request, @Nullable PropertyDefinition definition, Optional<ComponentDto> component) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: MINOR Specify a "ComponentDto" parameter instead. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues
* MAJOR 3 major
* MINOR 3 minor

Watch the comments in this conversation to review them.

**Comment by julienlancelot** [comment]
When the fieldValue is not a JSON, I've received a strange error :
```
"Ill-formatted JSON 'val1' for setting 'sonar.autogenerated'"
```

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java [hunk 12]
```diff
@@ -126,6 +146,64 @@ public void handle(Request request, Response response) throws Exception {
...
+
+    PropertyDefinition definition = propertyDefinitions.get(request.getKey());
+
+    if (!request.getFieldValues().isEmpty()) {
+      doHandlePropertySet(dbSession, request, definition, component);
+    } else {
+      validate(request, component);
+      dbClient.propertiesDao().insertProperty(dbSession, toProperty(request, component));
+      if (!component.isPresent()) {
+        settingsChangeNotifier.onGlobalPropertyChange(request.getKey(), persistedValue(request));
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: Why notification is not done when the setting is a property set ?
This is ugly, but I think we should return the value in DB (the list of generated ids of fields)

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java [hunk 13]
```diff
@@ -167,17 +256,25 @@ private void checkType(SetRequest request, PropertyDefinition definition) {
   }
 
   private static void checkValueIsSet(SetRequest request) {
-    checkRequest(isNullOrEmpty(request.getValue()) ^ request.getValues().isEmpty(),
-      "Either '%s' or '%s' must be provided, not both", PARAM_VALUE, PARAM_VALUES);
+    checkRequest(!isNullOrEmpty(request.getValue())
+      ^ !request.getValues().isEmpty()
+      ^ !request.getFieldValues().isEmpty(),
+      "Either '%s', '%s' or '%s' must be provided, not both", PARAM_VALUE, PARAM_VALUES, PARAM_FIELD_VALUES);
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: As there are 3 parameters, "not both" is not right

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java [hunk 14]
```diff
@@ -167,17 +256,25 @@ private void checkType(SetRequest request, PropertyDefinition definition) {
   }
 
   private static void checkValueIsSet(SetRequest request) {
-    checkRequest(isNullOrEmpty(request.getValue()) ^ request.getValues().isEmpty(),
-      "Either '%s' or '%s' must be provided, not both", PARAM_VALUE, PARAM_VALUES);
+    checkRequest(!isNullOrEmpty(request.getValue())
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: There's a missing check when none of these 3 properties is set.

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java [hunk 15]
```diff
@@ -112,12 +137,7 @@ public void define(WebService.NewController context) {
   public void handle(Request request, Response response) throws Exception {
     DbSession dbSession = dbClient.openSession(false);
     try {
-      SetRequest setRequest = toWsRequest(request);
-      Optional<ComponentDto> component = searchComponent(dbSession, setRequest);
-      checkPermissions(component);
-
-      validate(setRequest, component);
-      dbClient.propertiesDao().insertProperty(dbSession, toProperty(setRequest, component));
+      doHandle(dbSession, toWsRequest(request));
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: It's not obvious for me that the validation on value, values, fieldValues is done in the toWsRequest.
I would move ``checkValueIsSet(setRequest);`` here

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java [hunk 16]
```diff
@@ -203,6 +301,24 @@ private static SetRequest toWsRequest(Request request) {
...
+      .collect(Collectors.toList());
+  }
+
+  private static Map<String, String> readOneFieldValues(String json, String key) {
+    Type type = new TypeToken<Map<String, String>>() {
+    }.getType();
+    try {
+      return (Map<String, String>) GsonHelper.create().fromJson(json, type);
+    } catch (JsonParseException e) {
+      throw new BadRequestException(String.format("Ill-formatted JSON '%s' for setting '%s'", json, key));
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: As reported by my first comment, the error message is strange. Wha means "Ill-formated" ?

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java [hunk 17]
```diff
@@ -138,6 +216,17 @@ private void validate(SetRequest request, Optional<ComponentDto> component) {
     checkComponentQualifier(request, definition, component);
   }
 
+  private void checkFieldType(SetRequest request, @Nullable PropertyDefinition definition, ListMultimap<String, String> valuesByFieldKeys) {
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: Remove this useless Nullable annotation

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java [hunk 18]
```diff
@@ -126,6 +146,64 @@ public void handle(Request request, Response response) throws Exception {
...
+      settingsUpdater.deleteComponentSetting(dbSession, key, component.get());
+    } else {
+      settingsUpdater.deleteGlobalSetting(dbSession, key);
+    }
+  }
+
+  private void validatePropertySet(SetRequest request, @Nullable PropertyDefinition definition, Optional<ComponentDto> component) {
+    checkRequest(definition != null, "Setting '%s' is undefined", request.getKey());
+    checkRequest(PropertyType.PROPERTY_SET.equals(definition.type()), "Parameter '%s' is used for setting of property set type only", PARAM_FIELD_VALUES);
+    checkComponentQualifier(request, definition, component);
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: The checkGlobalOrProject() call is missing

Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: But maybe you could try to move this check at the begin of doHandle(), in order to not have to use it twice.

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java [hunk 19]
```diff
@@ -229,4 +343,18 @@ private PropertyDto toProperty(SetRequest request, Optional<ComponentDto> compon
 
     return property;
   }
+
+  private static PropertyDto toFieldProperty(KeyValue keyValue, @Nullable Long componentId) {
+    return new PropertyDto().setKey(keyValue.key).setValue(keyValue.value).setResourceId(componentId);
+  }
+
+  private static class KeyValue {
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: In my opinion, you should add more logic here : 
- The constructor should take the baseSettingKey, the fieldId, the Entry and the component
- The toFieldProperty could be moved inside

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java (hunk 18 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: Anyway, UT needs to be adding to cover the usage of property set settings only available on global | on some component qualifiers.

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java [hunk 20]
```diff
@@ -138,6 +216,17 @@ private void validate(SetRequest request, Optional<ComponentDto> component) {
     checkComponentQualifier(request, definition, component);
   }
 
+  private void checkFieldType(SetRequest request, @Nullable PropertyDefinition definition, ListMultimap<String, String> valuesByFieldKeys) {
+    for (PropertyFieldDefinition fieldDefinition : definition.fields()) {
+      for (String value : valuesByFieldKeys.get(fieldDefinition.key())) {
+        PropertyDefinition.Result result = fieldDefinition.validate(value);
+        checkRequest(result.isValid(),
+          i18n.message(Locale.ENGLISH, "property.error." + result.getErrorKey(), "Error when validating setting with key '%s'. Field '%s' has incorrect value '%s'."),
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: The message "Error when ..." is never displayed when bundle are not mocked (as in your UT).
For instance, here's the message I got : `` "Valid options are \"true\" and \"false\"`` => no idea which field was wrong.
I would suggest to not use i18n message when validating property set, and to always display your message "Error when...".

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java (hunk 13 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: Either one of them

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java (hunk 16 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: incorrect or invalid

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java (hunk 14 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: `One and only one of 'value', 'values', 'fieldValues' must be provided` address the issue

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java (hunk 19 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: not implemented but I agree on principle

**Pull request merged**